[@kbn/ui-framework] move ui-framework to a package

spalger opened the pull request:
Before opening x-pack we either need to remove our usage of react-router v3, which is used by the ui_framework doc site. Rather than make the massive amount of changes necessary to upgrade to react-router v4 this moves the ui_framework into a package so that it is isolated from core Kibana.

All yarn scripts have been updated to continue working as designed, the only real difference is that ui_framework components are now imported from `@kbn/ui-framework/components` rather than `ui_framework/components`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/487/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/490/)

spalger commented:
@cjcenizal can you please help me verify that the build tasks and such still work as expected?

spalger commented:
I'm also fairly confident that all the jest tests are still running, but if you could verify that would be appreciated.

spalger commented:
Got it, fixed and updated the .gitignore file to not ignore the index.html file (apparently you have to use the `*` notation for the negation to work properly).

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/537/)

kjbekkelund commented on .eslintrc.js:
```diff
@@ -41,5 +41,44 @@ module.exports = {
+        'import/no-extraneous-dependencies': [
+          'error',
+          {
+            devDependencies: true,
```
Why do we need this setup for `ui-framework` but not for the rest of Kibana? Maybe it's been discussed somewhere (I've been sick the last week, so my brain is still a bit foggy).

Or is it just a "nice addon" to help make it clear?

spalger replied:
I only added it to the ui-framework to help me verify that I was pulling in the right dependencies and making the right things dev/normal deps. I think we could absolutely use this in other places, but I don't think it needs to go into this PR.

kjbekkelund commented on package.json:
```diff
@@ -305,11 +292,8 @@
     "xml2js": "0.4.19",
-    "xmlbuilder": "9.0.4",
-    "yeoman-generator": "1.1.1",
-    "yo": "2.0.0"
```
I ❤️ THIS CLEANUP SO MUCH!

kjbekkelund commented on src/dev/jest/config.js:
```diff
@@ -44,15 +40,14 @@ export default {
   ],
   modulePathIgnorePatterns: [
     '__fixtures__/',
+    '<rootDir>/packages/kbn-ui-framework/target/',
```
Maybe we should just do `/target/` here? I felt I had already included that in some PR, but maybe it's one of my active ones 🙈

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/566/)

spalger commented:
6.x/6.3: 2013bb9
